File: 1. JPA basic/basic/src/main/java/study/jpa/basic/jpashop/domain/Order.java
package study.jpa.basic.jpashop.domain;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.EnumType;
import javax.persistence.Enumerated;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.persistence.Table;

/**
 * @author Joonhyuck Hyoung
 */
@Entity(name = "jpashop_order")
@Table(name = "jpashop_order")
public class Order extends BaseEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long no;

    @ManyToOne
    // 새로운 사실 JoinColumn 의 name
    // Member class의 pk를 넣어주는게 아니다. 애초에 @JoinColumn 어노테이션만으로 Member class의 pk를 읽는다.
    // 여기서 name은 이 Order클래스의 member 컬럼의 이름을 적어주는 것임
    /**
     *  create table jpashop_order (
     *        no bigint not null auto_increment,
     *         ordered_at datetime(6),
     *         status varchar(255),
     *         DELIVERY_ID bigint,
     *         MEMBER_ID bigint,
     *         primary key (no)
     *     ) engine=InnoDB
     */
    @JoinColumn(name = "MEMBER_ID")
    private Member member;

    @OneToOne
    @JoinColumn(name = "DELIVERY_ID")
    private Delivery delivery;

    @OneToMany(mappedBy = "order")
    private List<OrderItem> orderItems = new ArrayList<>();

    @Column(name = "ordered_at")
    private LocalDateTime orderedAt;

    @Enumerated(EnumType.STRING)
    private OrderStatus status;

    public Member getMember() {
        return member;
    }

    public void setMember(Member member) {
        this.member = member;
    }
    public Long getNo() {
        return no;
    }


    public LocalDateTime getOrderedAt() {
        return orderedAt;
    }

    public OrderStatus getStatus() {
        return status;
    }

    // 연관관계 편의 메서드
    public void addOrderItem(OrderItem orderItem) {
        orderItem.setOrder(this);
        orderItems.add(orderItem);
    }
}

File: 2. JPA deep/jpashop/src/main/kotlin/jpabook/jpashop/domain/Order.kt
package jpabook.jpashop.domain

import java.time.*
import javax.persistence.*

@Entity
@Table(name = "orders")
data class Order(
    @Id @GeneratedValue
    @Column(name = "order_id")
    val id: Long = 0L,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    private var member: Member,

    @OneToMany(mappedBy = "order", cascade = [CascadeType.ALL])
    val orderItems: MutableList<OrderItem> = arrayListOf(),

    @OneToOne(cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    @JoinColumn(name = "delivery_id")
    private var delivery: Delivery,

    val orderDate: LocalDateTime = LocalDateTime.now(), // 주문 시간

    @Enumerated(EnumType.STRING)
    var status: OrderStatus = OrderStatus.ORDER, // 주문 상세 (ORDER, CANCEL)
) {

    // == 연관관계 메서드 == //
    fun setMember(member: Member) {
        this.member = member;
        member.orders.add(this)
    }

    fun addOrderItem(orderItem: OrderItem) {
        orderItems.add(orderItem)
        orderItem.order = this
    }

    fun setDelivery(delivery: Delivery) {
        this.delivery = delivery
        delivery.order = this
    }

    //== 비즈니스 로직 ==//
    /** 주문 취소 **/
    fun cancel() {
        if (delivery.status == DeliveryStatus.COMP)
            throw IllegalStateException("이미 배송 완료된 상품은 취소가 불가능합니다.")

        status = OrderStatus.CANCEL
        orderItems.forEach { it.cancel() }
    }

    //== 조회 로직 ==//
    /** 전체 주문 가격 조회 **/
    fun getTotalPrice() = orderItems.sumOf {
        it.getTotalPrice()
    }

    //== 생성 메서드 ==//
    companion object {
        fun createOrder(member: Member, delivery: Delivery, orderItem: OrderItem): Order {
            return Order(
                member = member,
                delivery = delivery,
                status = OrderStatus.ORDER,
                orderDate = LocalDateTime.now(),
            ).also { order ->
                order.orderItems.add(orderItem)
            }
        }
    }
}
